website_sale_one_step_checkout/controllers/main.py

```diff
@@ -0,0 +1,225 @@
+        if post.get('xhr'):
+            return 'ok'
+
+        return request.render(
+            'website_sale_one_step_checkout.osc_onestepcheckout', values)
+
+    @http.route(['/shop/checkout/validate_address_form'], type='json',
+                auth='public', website=True, multilang=True)
+    def validate_address_form(self):
+        order = request.website.sale_get_order()
+
+        result = {
+            'success': False
+        }
+
+        if not order:
+            return result
+
+        if order.partner_id.id != request.website.user_id.sudo().partner_id.id:
+            for f in self._get_mandatory_billing_fields():
+                if not order.partner_id[f]:
+                    result['partner_id'] = order.partner_id.id
+                    return result
+
+            result['success'] = True
+
+            return result
+
+        elif order.partner_id.id == request.website.user_id\
+                .sudo().partner_id.id:
+            return result
+
+    @http.route()
+    def address(self, **post):
+        if not request.website.use_osc:
+            return super(WebsiteSale, self).address(**post)
+        else:
+            return request.redirect('/shop')
+
+    @http.route(['/shop/checkout/render_address'], type='json', auth='public',
+                website=True, multilang=True)
+    def render_address(self, **kw):
+        Partner = request.env['res.partner']\
+            .with_context(show_address=1).sudo()
+        order = request.website.sale_get_order(force_create=1)
+        def_country_id = order.partner_id.country_id
+        values, errors = {}, {}
+
+        mode = (False, False)
+        partner_id = int(kw.get('partner_id', -1))
+        shippings = []
+
+        redirection = self.checkout_redirection(order)
+        if redirection:
+            return redirection
+
+        # IF PUBLIC ORDER
+        if order.partner_id.id == request.website.user_id.sudo().partner_id.id:
+            mode = ('new', 'billing')
+            country_code = request.session['geoip'].get('country_code')
+            if country_code:
+                def_country_id = request.env['res.country'].search(
+                    [('code', '=', country_code)], limit=1)
+            else:
+                def_country_id = request.website.user_id.sudo().country_id
+        # IF ORDER LINKED TO A PARTNER
+        else:
+            if partner_id > 0:
+                if partner_id == order.partner_id.id:
+                    mode = ('edit', 'billing')
+                else:
+                    shippings = Partner.search([
+                        ('id', 'child_of',
+                         order.partner_id.commercial_partner_id.ids)],
+                        order='id desc')
+                    if partner_id in shippings.mapped('id'):
+                        mode = ('edit', 'shipping')
+                    else:
+                        return Forbidden()
+                if mode:
```

Comment:
Odoo standard: https://github.com/odoo/odoo/blob/10.0/addons/website_sale/controllers/main.py#L571

Reply:
I don't think it's a good argument if it's right way if it's in Odoo core.

Reply:
Actually I just realised the case where `mode` is not set does not occur. At this point, the `mode` will be either `('edit', 'billing')` or `('edit', 'shipping')`, otherwise `Forbidden()` would have been returned. So I think it's save to remove the `if mode:` and just move code to the left?